src/CRM/CivixBundle/Resources/Example/SchemaBuilderTrait.php

```diff
@@ -0,0 +1,256 @@
+        if (!count($xmlSchemas)) {
+          throw new \Exception("Could not find files matching '$xmlSchemaGlob'. You may want to run `civix generate:entity` before running this command.");
+        }
+
+        foreach ($xmlSchemas as $xmlSchema) {
+          $dom = new \DomDocument();
+          $xmlString = file_get_contents($xmlSchema);
+          $dom->loadXML($xmlString);
+          $xml = simplexml_import_dom($dom);
+          if (!$xml) {
+            throw new \Exception('There is an error in the XML for ' . $xmlSchema);
+          }
+          $this->specification->getTable($xml, $this->database, $this->tables);
+          $name = (string) $xml->name;
+          $this->tables[$name]['name'] = $name;
+          $sourcePath = substr($xmlSchema, strlen("{$this->basedir}/"));
+          $this->tables[$name]['sourceFile'] = $this->sourceFilePrefix . $sourcePath;
+        }
+
+        $this->orderTables($this->tables);
+        $this->resolveForeignKeys($this->tables);
+
+        return $this;
+      }
+
+      /**
+       * Create DAO files.
+       *
+       * @return $this
+       */
+      public function generateDaoFiles() {
+        $config = new \stdClass();
+        $config->phpCodePath = "{$this->basedir}/";
+        //    $config->sqlCodePath = $basedir->string('sql/');
+        $config->tables = $this->tables;
+
+        foreach ($this->tables as $table) {
+          $dao = new \CRM_Core_CodeGen_DAO($config, (string) $table['name'], $this->tsFunction);
+          // Don't display gencode's output
+          ob_start();
+          $dao->run();
+          ob_end_clean();
+          $file = $this->basedir . "/{$table['base']}{$table['fileName']}";
+          $this->files[] = ['type' => 'dao', 'file' => $file];
+        }
+
+        return $this;
+      }
+
+      /**
+       * @param string $type
+       *   The type of SQL file to generate, 'CREATE' or 'DROP'.
+       *
+       * @return string
+       * @throws \Exception
+       */
+      public function generateSql($type) {
+        $generators = ['CREATE' => 'generateCreateSql', 'DROP' => 'generateDropSql'];
+        $generator = $generators[strtoupper($type)] ?? NULL;
+        if (!$generator) {
+          throw new \Exception("Invalid SQL content type: $type");
+        }
+
+        $tmpFile = sys_get_temp_dir() . DIRECTORY_SEPARATOR . bin2hex(random_bytes(12));
+        $config = new \stdClass();
+        $config->sqlCodePath = dirname($tmpFile);
+        $config->tables = $this->tables;
+
+        $schema = new \CRM_Core_CodeGen_Schema($config);
+
+        // We're poking into an internal class+function (`$schema->$generator()`) that changed in v5.23.
+        // Beginning in 5.23: $schema->$function() returns an array with file content.
+        // Before 5.23: $schema->$function($fileName) creates $fileName and returns void.
+        if (version_compare(\CRM_Utils_System::version(), '5.23.alpha1', '>=')) {
+          return array_shift($schema->$generator());
+        }
```

Comment:
I was playing with this trait a bit to see if I could use it standalone. Partial success but this line didn't work and it seems like it should always fail?? `array_shift()` uses pass by reference, so it needs an actual variable.

The main other thing that came up is there's no way to set $database and it doesn't seem to set it internally either, so it errors in a couple places.